torch/csrc/autograd/variable.h

```diff
@@ -449,17 +449,28 @@ struct TORCH_API DifferentiableViewMeta : public AutogradMeta {
 // Differentiable view. Track history with DifferentiableViewMeta.
 inline Variable make_variable_differentiable_view(
     Variable base,
-    at::Tensor data,
+    at::Tensor& data,
     CreationMeta creation_meta,
     c10::optional<std::function<at::Tensor(const at::Tensor&)>> view_func = c10::nullopt) {
   if (data.defined()) {
-    auto data_impl_copy = data.getIntrusivePtr()->shallow_copy_and_detach(
-      /*version_counter=*/0,
-      /*allow_tensor_metadata_change=*/true);
-    data_impl_copy->set_autograd_meta(std::make_unique<DifferentiableViewMeta>(
-      data_impl_copy.get(), std::move(base), std::move(view_func),
-      creation_meta));
-    return Variable(data_impl_copy);
+    // If we already did a TensorImpl allocation for data, just reuse it.
+    // Otherwise(e.g tensor.swapdim(0, 0) when we return the same tensor as input),
+    // we have to use shallow_copy_and_detach to create a new TensorImpl to avoid
+    // moving leaf node into graph interior. This guarantees only 1 TensorImpl
+    // allocation happens in view ops.
+    if (data.getIntrusivePtr().use_count() == 1 && data.getIntrusivePtr()->unique_version()) {
+      auto data_impl = data.unsafeGetTensorImpl();
+      data_impl->set_autograd_meta(std::make_unique<DifferentiableViewMeta>(
+        data_impl, std::move(base), std::move(view_func), creation_meta));
+      return Variable(data);
+    } else {
+      auto data_impl = data.getIntrusivePtr()->shallow_copy_and_detach(
```

Comment:
The naming here is pretty skeevy, because the type of `data_impl` in the two branches is different. In the top branch, it's non-owning, but this branch, it is owning, that is pretty confusing.